sid/visualize_simulation_results.py

```diff
@@ -2,249 +2,227 @@
+    for var, color in zip(vars_to_plot, colors):
+        y_range = (0, 1) if var in full_range_vars else None
+        bg_values = rates[var].columns.unique().levels[0]
+        for bg_val in bg_values:
+            plot_title = f"{_nice_str(var)} {title}"
+            if bg_val != "general":
+                plot_title += f": {bg_val}"  # noqa
+            p = _plot_rates(
+                rates=rates[var][bg_val],
+                plot_title=plot_title,
+                color=color,
+                y_range=y_range,
+            )
+            p.name = var if bg_val == "general" else f"{var}_{bg_val.replace(' ', '')}"
+            plots.append(p)
+    return plots
+
+
+def _plot_rates(rates, title, color, y_range):
     """Plot the rates over time.
 
     Args:
-        means (DataFrame):
-            the index are the x values, the values the y values.
-            every column is plotted as a separate line.
-        colors (list): full color palette
+        rates (DataFrame): the index are the x values, the values the y values.
+            Every column is plotted as a separate line.
+        color (str): color.
         title (str): plot title.
+        y_range (tuple or None): range of the y axis.
+
+    Returns:
+        p (bokeh figure)
 
     """
-    p = figure(tools=[], plot_height=300, plot_width=600, title=title,)
-    for var, color in zip(means, colors[: len(means.columns)]):
-        rates = means[var]
-        p.line(
-            x=rates.index, y=rates, line_width=2.0, legend_label=var, line_color=color,
-        )
-    if means.columns.name is not None:
-        p.legend.title = _nice_str(means.columns.name)
+    xs = rates.index
+    p = figure(
+        tools=[],
+        plot_height=400,
+        plot_width=800,
+        title=title,
+        y_range=y_range,
+        x_axis_type="datetime",
+    )
+
+    # plot the median
+    p.line(x=xs, y=rates.median(axis=1), alpha=1, line_width=2.75, line_color=color)
+
+    # plot the confidence band
+    q5 = rates.apply(np.nanpercentile, q=5, axis=1)
+    q95 = rates.apply(np.nanpercentile, q=95, axis=1)
+    p.varea(x=xs, y1=q95, y2=q5, alpha=0.2, color=color)
+
+    # add the trajectories
+    for var in rates:
+        p.line(x=xs, y=rates[var], line_width=1, line_color=color, alpha=0.3)
+
     p = _style(p)
     return p
 
 
+def _export_plots_and_layout(title, plots, outdir_path):
+    """Save all plots as png and the layout as html.
+
+    Args:
+        title (bokeh.Div): title element.
+        plots (list): list of bokeh plots
+        outdir_path (pathlib.Path): base path to which to append the plot name to build
+            the path where to save each plot.
+
+    """
+    for p in plots:
+        outpath = outdir_path / f"{p.name}.png"
+        output_file(outpath)
+        export_png(p, filename=outpath)
+
+    output_file(outdir_path / "overview.html")
+    save(Column(title, *plots))
+
+
+def _input_processing(data_paths, outdir_path, background_vars):
```

Comment:
This function does too much and is used only in one place, so we do not avoid any repititions if we break it apart.

1. I would handle colors and background_vars at the top of `visualize...()`.
2. I would use one function to process the input data. Handle dataframes and paths equally by creating a list. Then, we need one case distinction where you use `pd.read_pickle` depending on whether you get a path or a dataframe.
3. Create folders stays the same.

Reply:
Yes, thanks for the suggestion. I did that in 4c2cf5bd53431ac08f6d394c896e385a341c9728 and that makes it easier to see what's happening.